Add max upstream connections dns-proxy option

dvejmz opened the pull request:
Allows defining a limit to the number of connections that can be established with the upstream DNS host. If left unset, there may be situations where connections fail to establish, which causes the Transport to create an influx of connections causing upstream to throttle our requests and triggering a runaway effect resulting in high CPU usage.

This is a workaround for [the unresolved upstream connection issues](https://github.com/cloudflare/cloudflared/issues/23#issuecomment-687743790) that some people have reported occurring in Raspberry Pis running Pi-hole. By allowing users to cap the number of connections themselves to adjust them to the capacity of their specific device, we're at least mitigating against the impact this runaway effect has on the limited CPU resources these devices have.

```
Jan 09 17:38:46 hostname cloudflared[9896]: 5:38PM ERR failed to connect to an HTTPS backend "https://1.0.0.1/dns-query" error="failed to perform an HTTPS request: Post \"https://1.0.0.1/dns-query\": context deadline exceeded"
```

I'm currently running a cloudflared binary compiled with this changeset like so

```
cloudflared proxy-dns --port 54 --upstream https://1.1.1.1/dns-query --upstream https://1.0.0.1/dns-query --max-upstream-conns 5
```

Which brings down the %CPU of the cloudflared task in my Raspberry Pi 1 Model B from 80-98% to the 0-5% range.

This fix is based on @kitsook's [changeset](https://github.com/kitsook/cloudflared/commit/134410ae07e559743d50ee516b8b4a5111e01200). See #91 for more details about the issue.

adamchalmers commented:
Great, thanks very much for this PR! @sudarshan-reddy reproduced and confirmed this fixes it.

adamchalmers commented on cmd/cloudflared/tunnel/server.go:
```diff
@@ -13,7 +13,11 @@ func runDNSProxyServer(c *cli.Context, dnsReadySignal, shutdownC chan struct{},
 		return errors.New("The 'proxy-dns-port' must be a valid port number in <1, 65535> range.")
 	}
-	listener, err := tunneldns.CreateListener(c.String("proxy-dns-address"), uint16(port), c.StringSlice("proxy-dns-upstream"), c.StringSlice("proxy-dns-bootstrap"), log)
+	maxUpstreamConnections := c.Int("proxy-dns-max-upstream-conns")
```
Shouldn't this be `c.Int("max-upstream-conns")` to match the [definition](https://github.com/cloudflare/cloudflared/pull/290/files?diff=unified&w=1#diff-6f1cb455303a0da3ac60ed8b8cdcb845a5edb82939f56e2e05939cb625522a76)

adamchalmers commented on tunneldns/tunnel.go:
```diff
@@ -68,6 +68,12 @@ func Command(hidden bool) *cli.Command {
+			&cli.IntFlag{
+				Name:    "max-upstream-conns",
+				Usage:   "Maximum concurrent connections to upstream, unlimited by default",
+				Value:   0,
```
I think 0 is a dangerous default here, because it'll cause the bad behaviour. Let's make the default a positive number. I'm not sure what is appropriate here... 10? Maybe 100? And let's put it into a `const` so that other code can reference the default too (e.g. `func configureBootstrap` and `func MaxUpstreamConnectionsOrDefault`).

adamchalmers commented:
@dvejmz @sudarshan-reddy I pushed up a new commit, could you let me know what you think?

sudarshan-reddy commented on cmd/cloudflared/config/model.go:
```diff
@@ -25,11 +27,12 @@ type Tunnel struct {
+	Port                   uint16   `json:"port,omitempty"`
+	Upstreams              []string `json:"upstreams,omitempty"`
+	Bootstraps             []string `json:"bootstraps,omitempty"`
+	MaxUpstreamConnections int      `json:"max_upstream_connections,omitempty"`
```
I wonder if this setting belongs here: https://github.com/cloudflare/cloudflared/blob/master/cmd/cloudflared/tunnel/cmd.go#L985 instead.

ipostelnik replied:
Do we really need this to be configurable? How would the users know how to set this? Can we pick a reasonable default and go with it? Say 100 connections or even 1000?

sudarshan-reddy replied:
I like your point. We could set that default here: https://github.com/cloudflare/cloudflared/pull/290/commits/c2f51021199b57f7d7f104526208b52763dc6984#diff-6f1cb455303a0da3ac60ed8b8cdcb845a5edb82939f56e2e05939cb625522a76R27 and still keep it override-able if anyone should choose to want a different number (even 0).

adamchalmers replied:
We should probably never let the user choose 0 as it will crash their machine by eating 100% CPU, so yeah, maybe Igor's right and this doesn't need to be configurable.

kitsook replied:
I think having it configurable makes sense. Say Raspberry Pi may see 100% CPU with 10 concurrent connections. But an enterprise server may easily handle 50 connections.

If there is concern with unlimited upper limit, maybe set the min as 1 and disallow setting it to 0?

As for the default value, it should be something that won't trigger throttling with most DoH providers. Say if the provider has a limit of 5 and the default max is 10, in the worst case, the provide may just reject requests from all 10 connections while cloudflared keeps re-creating 10 connections in the pool.

dvejmz replied:
Thanks all for your comments on this. I will review your suggestions this evening and update the PR. I do agree providing a default value would save the user the trial-and-error process involved in tweaking it to their hardware. For me, a value of 5 upstream connections gives me really good results on a RPi 1 Model B, but it may be possible to increase it if necessary on multi-core devices.

adamchalmers replied:
Default of 5 sounds good, at least we know it works for some users. And let's keep it configurable, it should be able to run well on any hardware.

sudarshan-reddy reviewed (commented):
The revision LGTM. I just have one question.

dvejmz commented:
I have pushed some changes addressing your feedback:

- I've appended the `max-upstream-conns` CLI flag to `tunneldns/tunnel.go` as per @sudarshan-reddy 's suggestion. I have not removed it from `cmd/cloudflared/config/model.go` as I think it's still required there so that we can initialise the `tunneldns` listener within the `ResolverService` wrapper. Please let me know if there's something I misunderstood in your feedback, I'll be happy to change it.
- I've reduced the default max upstream connections from `10` to `5` @adamchalmers . I've not been able to reproduce the issue today on my Raspberry Pi so I didn't get a chance to verify if `10` would actually be a safe default as well. I know a value of `5` is guaranteed to work as I did confirm it performed correctly when connections to the HTTPS upstreams fail. Further testing in the future may lead to it being increased to 10 (or higher) in the future. I suppose as long as it remains configurable, it shouldn't be an issue either way.

I hope that makes sense. I recompiled and reinstalled this newer patched version in my RPi and everything seems to be working nicely. Do let me know if you have any other comments!

ipostelnik commented on cmd/cloudflared/tunnel/cmd.go:
```diff
@@ -1003,6 +1003,13 @@ func configureProxyDNSFlags(shouldHide bool) []cli.Flag {
 			Hidden:  shouldHide,
 		}),
+		altsrc.NewIntFlag(&cli.IntFlag{
+			Name:    tunneldns.MaxUpstreamConnsFlag,
```
You can't use `max-upstream-conns` as flag name here, it's not clear that it's used for tunneling DNS only. You need to prefix it with `proxy-dns-`. You'll need to change the code that reads it as well.

dvejmz replied:
Good spot, will fix, thanks!

ipostelnik commented:
@dvejmz thanks for implementing these changes

Stephen988 commented:
Hi @dvejmz can you explain step by step how to fix cloudflared? I don't understand where apply this fix for reduce max connections.
Thanks!

dvejmz commented:
HI @Stephen988 ! In order to apply this fix, make sure you've downloaded the latest cloudflared binary (or at least any version equal or later than [2021.2.3](https://github.com/cloudflare/cloudflared/releases/tag/2021.2.3)). If you cannot find the right binary for your machine/architecture ─ if you're running it in an older Raspberry Pi like a Pi 1 ─ you may need to compile `cloudflared` yourself. To do that, make sure you've got `go` installed and then build it from the git repo with a command similar to the one below, substituting your target OS and architecture in the env vars below. The build config below will build it for a Raspberry Pi 1.

```
GOOS=linux GOARCH=arm GOARM=6 go build -v "-ldflags=-X 'main.Version=dev' -X 'main.BuildTime=$(date)'" ./cmd/cloudflared
```

Once you've got the cloudflared binary in the Pi, you can launch it as normal, as it should work fine with the current default max connections parameter. If you wish to change this parameter regardless, you'll need to pass it as an option when launching `cloudflared` as below

```
cloudflared proxy-dns [other flags you may be already using here] --max-upstream-conns <desired_number>
```

Hope that helps.

Stephen988 commented:
Hi @dvejmz I'm very noob, never compiled a software, I have a raspberry pi 4 running raspbian os 32bit, actually I have the most recent version of docker, portainer and I'm using crazymax/cloudflared:latest.
how and where insert your string? sorry but is a new field for me.

kitsook commented:
@Stephen988 You can set the parameter via environment variable, e.g.
```
docker run ... --env TUNNEL_DNS_MAX_UPSTREAM_CONNS=<desired_number> ...
```

Stephen988 commented:
> @Stephen988 You can set the parameter via environment variable, e.g.
> 
> ```
> docker run ... --env TUNNEL_DNS_MAX_UPSTREAM_CONNS=<desired_number> ...
> ```

nothing changes, always error: ERR failed to connect to an HTTPS backend "https://1.1.1.1/dns-query" error="failed to perform an HTTPS request: Post \"https://1.1.1.1/dns-query\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
